kernel/src/benchmarks/oqueue.rs

```diff
@@ -0,0 +1,831 @@
+    }
+
+    /// Get the value present in this Slot. This is only safe if [`Slot::store`] was previously
+    /// called.
+    unsafe fn get(&self) -> T {
+        // SAFETY: This assumes that the self.store has be previously called
+        unsafe { ptr::read(self.value.data.get()).assume_init() }
+    }
+
+    #[expect(unused)]
+    unsafe fn get_maybe_uninit(&self) -> MaybeUninit<T> {
+        // SAFETY: This assumes that the self.store has be previously called
+        unsafe {
+            let data = ptr::read(self.value.data.get());
+            data
+        }
+    }
+}
+
+fn allocate_page_aligned_array<T>(len: usize) -> Box<[Slot<T>]> {
+    let elem_size = core::mem::size_of::<Slot<T>>();
+    let total_size = elem_size * len;
+
+    // Layout with page alignment
+    let layout = core::alloc::Layout::from_size_align(total_size, 4096).expect("invalid layout");
+
+    unsafe {
+        let ptr = alloc(layout);
+        if ptr.is_null() {
+            handle_alloc_error(layout);
+        }
+
+        // Initialize each element
+        let mut current = ptr as *mut Slot<T>;
+        for _ in 0..len {
+            ptr::write(current, Slot::new());
+            current = current.add(1);
+        }
+
+        // Turn into a Box<[Slot<T>]> safely
+        let slice = core::slice::from_raw_parts_mut(ptr as *mut Slot<T>, len);
+        Box::from_raw(slice)
+    }
+}
+
+pub struct Rigtorp<T> {
+    this: Weak<Self>,
+    capacity: NonZero<usize>,
+    slots: Box<[Slot<T>]>,
+    head: CachePadded<AtomicUsize>,
+    tail: CachePadded<AtomicUsize>,
+}
+
+unsafe impl<T> Sync for Rigtorp<T> {}
+unsafe impl<T> Send for Rigtorp<T> {}
+
+impl<T> Rigtorp<T> {
+    /// new
+    pub fn new(size: usize) -> Arc<Self> {
+        assert!(size.is_power_of_two());
+        assert!(size > 0);
+
+        Arc::new_cyclic(|this| Rigtorp {
+            this: this.clone(),
+            capacity: NonZero::new(size).unwrap(),
+            slots: allocate_page_aligned_array(size),
+            head: CachePadded::new(AtomicUsize::new(0)),
+            tail: CachePadded::new(AtomicUsize::new(0)),
+        })
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity.into()
+    }
+
+    /// "turn" is similar to a generation counter.
+    /// N.B.: The way rigtorp/MPMCQueue uses the term turn here is misleading. `Slot`s have a turn,
+    /// but this is really the generation. The actual "turn" is either (2 * self.turn(index)) or
+    /// (2 * self.turn(index) + 1) for producers and consumers respectively. If I could rename it,
+    /// it would be "generation", but this ensures consistency with the rigtorp implementation to
+    /// make it easier for new readers to pick up.
+    fn turn(&self, index: usize) -> usize {
+        index / self.capacity
+    }
+
+    /// The physical index in the ringbuffer.
+    fn idx(&self, index: usize) -> usize {
+        index % self.capacity
+    }
+
+    /// produce
+    pub fn produce(&self, data: T)
```

Comment:
I'm assuming this is to silence warnings. In the future I would rather have `#[expect(missing_docs)]`. That makes sure there is a marker that there is a warning that by rights should be flagged, but is being ignored.